### DISCUSSION TITLE
Caddy / Reverse Proxy for Rustdesk server

### DISCUSSION THREAD

**Discussion by kylebial0** [rustdesk/rustdesk-server#427] [Q&A]
Is there any known way of hosting the rustdesk server behind a reverse proxy?

I have done a good few hours of testing and research, and from what I can tell, there doesn't appear to be any good way of doing so. 

My reason for wanting to do this is because I am looking to host the rustdesk server on my homelab behind a Caddy reverse proxy, mostly because I do not want to expose the additional ports required natively by Rustdesk. 

I look forward to hearing the communities thoughts on this topic, and to see if it has been successfully accomplished in the past.

**Comment by alexipeck** [comment]
I've gone down the same rabbit hole, the primary limitation I've found is Caddy not natively supporting UDP required by the HBBS (ID Server) binary. There's a project which could provide a solution (https://github.com/mholt/caddy-l4). I haven't managed to get it to work yet :(

Nginx should be capable of this.

**Comment by xlionjuan** [comment]
You could use Traefik, but unless you implement IP ACL, reverse proxy **WON'T** increase security.

**Comment by tschettervictor** [comment]
Simply build caddy with this plugin which allows layer 4 proxying using the caddyfile. Here is a snippet from one of my configs. This would be the global block.

```
{
	layer4 {
		tcp/:8189 {
			proxy {
				to 192.168.1.171:8189
			}
		}
	}
	email my email
}
```

**Reply by oswaldo-be** [reply]
So this should be the head of my Caddyfile?
and the proxy should be the IP of the machine running rustdesk?
so the config of rustdesk should look like this?

rustd.my.domain {
        reverse_proxy /api/* RUSTDESK-IP:21118
        reverse_proxy / {
                to RUSTDESK-IP:21119
                header_up Host {http.request.host}
                header_up X-Real-IP {http.request.remote}
                # Web-Socket-Unterstützung
                header_up Connection {http.request.header.Connection}
                header_up Upgrade {http.request.header.Upgrade}
        }
}

**Reply by xlionjuan** [reply]
You don't need any additional WebSocket or realIP settings for caddy, it will do it automatically.

**Reply by xlionjuan** [reply]
This is the sample config for Caddy, the WebSocket is working for Web/Native Client, API server didn't valid, but should working, for RustDesk Server's raw TCP/UDP port (21115 ~ 21117), I **do not** recommend to put reverse proxy in front of it.

Please replace `172.18.0.1` to your Caddy's Docker network gateway, or use `127.0.0.1` if you installed Caddy to host system, I recommend to install Caddy via container.

For `header_common`, it should be great for most people, but you could modify if needed.
```
(header_common) {
	header {
		# HSTS, very recommend to enable, for details, check https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Strict-Transport-Security
		# Strict-Transport-Security "max-age=31536000; includeSubDomains; preload"
		X-Content-Type-Options "nosniff"
		Referrer-Policy "same-origin"
		Content-Security-Policy "base-uri 'self';"
		X-Frame-Options "SAMEORIGIN"
		X-XSS-Protection "1; mode=block"
		# Remove reverse proxy header
		-Server
		-via
	}
}

(cors) {
	@origin header Origin {args[0]}
	header @origin Access-Control-Allow-Origin "{args[0]}"
	header @origin Access-Control-Allow-Methods "OPTIONS,GET,POST"
}

rustdesk.example.com {
	# Put reverse proxy to same route to import shared header
	route {
		import header_common
		import cors rustdesk.com
		encode zstd gzip

		# >>> Only for Pro Server
		# reverse_proxy / 172.18.0.1:21114 {}
		# <<< Only for Pro Server

		# ID Server
		reverse_proxy /ws/id http://172.18.0.1:21118 {
			transport http {
				read_timeout 120s
			}
		}

		# Relay Server
		reverse_proxy /ws/relay http://172.18.0.1:21119 {
			transport http {
				read_timeout 120s
			}
		}
	}
}
```
